### PR TITLE
[Bugfix] Fix MarlinFP8 weight transpose silently skipped for square matrices (N==K)

### LINKED ISSUE #44110: [Bug] MarlinFP8 kernel silently skips weight transpose for square matrices (N==K), corrupting FP8 inference on sm_75–sm_88 GPUs
## Environment

- **GPU**: sm_75–sm_88 (e.g. A10, A40, RTX A6000) — any GPU where `supports_fp8()` returns `False`
- **Model**: any FP8 model with square weight matrices, e.g. `ibm-granite/granite-4.1-8b-fp8`
- **Quantization**: `compressed-tensors` or `modelopt` FP8 with channel/tensor-wise scales

## Symptoms

FP8 models produce garbage output (repeated punctuation, incoherent tokens) on GPUs lacking native FP8 compute, while the identical non-FP8 model works correctly on the same hardware. No errors, no NaN — silent data corruption.

```
# FP8 model on A40 (sm_86)
OUTPUT: ',,,,,,,,,,,,,,,,,,,,'

# BF16 model on A40 (sm_86)
OUTPUT: 'The capital of France is Paris...'
```

## Root Cause

`MarlinFP8ScaledMMLinearKernel.process_weights_after_loading` uses a shape comparison to decide whether to transpose the weight to `(K, N)` before passing it to the Marlin kernel:

```python
# vllm/model_executor/kernels/linear/scaled_mm/marlin.py
if w_q.shape != (layer.input_size_per_partition, layer.output_size_per_partition):
    replace_parameter(layer, "weight", w_q.t())
```

For **square matrices where `N == K`** (e.g. `q_proj` and `o_proj` where `hidden_dim × hidden_dim = 4096×4096`), the tuples `(N, K)` and `(K, N)` are **identical**. The condition always evaluates `False` and the transpose is silently skipped. The weight stays in `(N, K)` checkpoint orientation when Marlin requires `(K, N)`, producing completely wrong GEMM results.

In dense models, `q_proj` and `o_proj` are always square. Their corrupted outputs overwhelm the residual stream and cascade through all layers.

## Why it escaped CI

- Most test shapes are non-square (e.g. MLP `gate_proj` at `12800×4096`) — those always transposed correctly.
- sm_89+ GPUs (H100, H200) use the native FP8 GEMM path and never reach the Marlin branch.
- Isolated Marlin kernel tests all pass — the kernel math is correct; only the input weight layout is wrong.

## Affected Code Path

On sm_75–sm_88, FP8 models fall back through:
- **CompressedTensors**: `CompressedTensorsW8A8Fp8` (needs sm_89) → `CompressedTensorsW8A16Fp8` → `MarlinFP8ScaledMMLinearKernel`
- **ModelOpt**: `ModelOptFp8LinearMethod` / `ModelOptFp8PcPtLinearMethod` → same `MarlinFP8ScaledMMLinearKernel`

## Related

- PR #38092 — introduced the regression
- Issue #33314 — layout canonicalization (referenced in the TODO comment in the affected code)

### PR DESCRIPTION
**Pull request by shernshiou**
## Purpose

Fix silent data corruption in `MarlinFP8ScaledMMLinearKernel` on GPUs without native FP8 compute (sm_75–sm_88). Fixes #44110.

### Root Cause

`process_weights_after_loading` used a shape comparison to decide whether to transpose the weight to `(K, N)` before passing it to Marlin:

```python
if w_q.shape != (layer.input_size_per_partition, layer.output_size_per_partition):
    replace_parameter(layer, "weight", w_q.t())
```

For **square matrices where `N == K`**, `(N, K)` and `(K, N)` are identical tuples — the check always evaluates `False` and the transpose is silently skipped. This was introduced in PR #38092, which also added `.t()` pre-transposes in two modelopt callers as a workaround, but that doesn't fix the CompressedTensors path and still fails for square shapes.

### Full blast radius

Three independent callers all hand `(N, K)` weights to `MarlinFP8ScaledMMLinearKernel.process_weights_after_loading`, and all were silently corrupt for square layers before this fix:

| Caller | Path |
|--------|------|
| `CompressedTensorsW8A16Fp8` | `create_fp8_weight_parameter` → `(N, K)`, no pre-transpose |
| `Fp8LinearMethod` (`use_marlin=True`) | `ops.scaled_fp8_quant(layer.weight)` returns `(N, K)`, no pre-transpose (see `fp8.py:553`) |
| `ModelOptFp8LinearMethod` / `ModelOptFp8PcPtLinearMethod` | pre-transpose added by #38092 as a workaround, removed here |

The `Fp8LinearMethod` non-Marlin `else` branch at `fp8.py:558` explicitly does `qweight.t()` itself — confirming the intended invariant was always "caller transposes before non-Marlin kernels, Marlin handles it internally", but the Marlin internal handling was broken for square shapes.

### Unit test

New test: `tests/kernels/quantization/test_marlin_fp8_kernel.py`

Exercises `process_weights_after_loading` + `apply_weights` end-to-end for both square (`N==K`) and non-square shapes, comparing against a BF16 reference. This is the exact scenario that triggered the regression — no existing test covered the full kernel path with square weights.

```bash
VLLM_TEST_FORCE_FP8_MARLIN=1 pytest \
    tests/kernels/quantization/test_marlin_fp8_kernel.py -v
```

### End-to-end test

Compared output of `ibm-granite/granite-4.1-8b` (BF16) against `ibm-granite/granite-4.1-8b-fp8` on an A40 (sm_86, Ampere) — a GPU that falls back to the Marlin path since it lacks native FP8 compute:

```bash
python -c "
from vllm import LLM, SamplingParams
params = SamplingParams(max_tokens=20)
prompt = ['Q: What is the capital of France? A:']

llm_bf16 = LLM('ibm-granite/granite-4.1-8b')
print('BF16:', llm_bf16.generate(prompt, params)[0].outputs[0].text)

llm_fp8 = LLM('ibm-granite/granite-4.1-8b-fp8', enforce_eager=True)
print('FP8: ', llm_fp8.generate(prompt, params)[0].outputs[0].text)
"
```

## Test Result

**Before fix** (A40, sm_86):
```
BF16: The capital of France is Paris...
FP8:  ,,,,,,,,,,,,,,,,,,,,
```

**After fix** (A40, sm_86):
```
BF16: The capital of France is Paris...
FP8:  The capital of France is Paris. Paris is known for its historical landmarks...
```

All 24 unit test cases pass (6 shapes × 2 dtypes).

### CODE REVIEW

**Comment by hclsys** [comment]
nice catch — the unconditional-transpose invariant is cleaner than a `(N==K)` special-case.

one thing not enumerated in the body: `Fp8LinearMethod.process_weights_after_loading` at `vllm/model_executor/layers/quantization/fp8.py:552` also calls `self.fp8_linear.process_weights_after_loading(layer)` on the `use_marlin=True` branch without any pre-`.t()`. by your invariant this is fine because `ops.scaled_fp8_quant` returns `(N, K)` and marlin now always-transposes — but it means the online-quant path (square attention projections, FP8 model, sm_75–88) was ALSO silently corrupt before your fix and is fixed by the same change. worth a line in the PR body so reviewers see the full blast radius.

any Ampere CI test in `tests/quantization/test_fp8.py` you'd like added to cover the square-shape regression? `q_proj` on a 4096-dim model is a one-line repro.

File: vllm/model_executor/layers/quantization/modelopt.py [hunk 1]
```diff
@@ -516,7 +516,7 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
             max_w_scale, weight = requantize_with_max_scale(
                 layer.weight, layer.weight_scale, layer.logical_widths
             )
-        layer.weight = Parameter(weight.t(), requires_grad=False)
+        layer.weight = Parameter(weight, requires_grad=False)
```
[the remark this answers is not in the document]
Comment (shernshiou) [reply vllm/model_executor/layers/quantization/modelopt.py]: Good catch — you're right. Removing `.t()` here breaks non-Marlin paths on sm_89+ because the base `FP8ScaledMMLinearKernel.process_weights_after_loading` is a no-op and `apply_weights` derives the output dim from `w.shape[1]`, expecting `(K, N)`.

Fixed in the latest commit: `.t()` is restored at this line so `modelopt` callers pass `(K, N)` non-contiguous into the kernel. `marlin.py` now uses `w_q.is_contiguous()` as the discriminator instead of shape equality — checkpoint weights are always contiguous `(N, K)` → transpose; a `.t()` view is always non-contiguous `(K, N)` → no-op. This correctly handles square matrices (N == K) which `.is_contiguous()` distinguishes even when shapes are identical.

File: vllm/model_executor/layers/quantization/modelopt.py [hunk 2]
```diff
@@ -598,7 +598,7 @@ def create_weights(
         )
 
     def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
-        layer.weight = Parameter(layer.weight.t(), requires_grad=False)
+        layer.weight = Parameter(layer.weight.data, requires_grad=False)
```
[the remark this answers is not in the document]
Comment (shernshiou) [reply vllm/model_executor/layers/quantization/modelopt.py]: Same fix applied here — `.t()` restored so `ModelOptFp8PcPtLinearMethod` passes `(K, N)` non-contiguous, which `marlin.py`'s `is_contiguous()` check correctly identifies as already transposed.

**Comment by shernshiou** [comment]
Thanks @hclsys! The square-shape regression is covered in `tests/kernels/quantization/test_marlin_fp8_kernel.py` with explicit 4096×4096 shapes (matching `q_proj`/`o_proj` in a 4096-dim model) for both the direct checkpoint path and the pre-transposed modelopt path. The `depthfirst-app` review comments have also been addressed in the latest commit — see replies inline.

**Comment by shernshiou** [comment]
cc @tjtanaa

**Comment by hclsys** [comment]
thanks @shernshiou — re my fp8.py:552 point: confirmed fp8.py just delegates to `self.fp8_linear.process_weights_after_loading` (which is the MarlinFP8 kernel when `use_marlin`), so your fix in marlin.py covers that path. 4096² in the new tests hits the square-shape regression directly. lgtm.

**Review by tjtanaa — APPROVED** [review]
LGTM

**Review by mgoin — CHANGES_REQUESTED** [review]
I believe this patches over the core issue, which is that some quant methods aren't preparing their weights into the same "canonical" format that the kernel expects. I'll look into fixing it at the source rather than at the kernel level
